roscala/src/main/scala/coop/rchain/rosette/prim/queue.scala

```diff
@@ -0,0 +1,196 @@
+    override val name: String = "queue-empty?"
+    override val minArgs: Int = 1
+    override val maxArgs: Int = 1
+
+    @checkTypeMismatch[Queue]
+    @checkArgumentMismatch
+    override def fn(ctxt: Ctxt): Either[PrimError, RblBool] = {
+      val q = ctxt.argvec.elem.head.asInstanceOf[Queue]
+      Right(RblBool(q.isEmpty()))
+    }
+  }
+
+  object queueEnqueue extends Prim {
+    override val name: String = "queue-enqueue"
+    override val minArgs: Int = 2
+    override val maxArgs: Int = 2
+
+    @checkArgumentMismatch
+    override def fn(ctxt: Ctxt): Either[PrimError, Queue] =
+      ctxt.argvec.elem.head match {
+        case q: Queue =>
+          val ob = ctxt.argvec.elem(1)
+          Right(q.enqueue(ob))
+        case _ =>
+          Left(ArgumentMismatch("The first element should be a Queue"))
+      }
+  }
+
+  object queueDeQueue extends Prim {
+    override val name: String = "queue-dequeue"
+    override val minArgs: Int = 1
+    override val maxArgs: Int = 1
+
+    @checkArgumentMismatch
+    override def fn(ctxt: Ctxt): Either[PrimError, Ob] =
+      ctxt.argvec.elem.head match {
+        case q: Queue =>
+          if (q.isEmpty()) {
+            Left(QueueEmptyError)
+          } else {
+            Right(q.dequeue().get)
+          }
+        case _ =>
+          Left(ArgumentMismatch("The first element should be a Queue"))
+      }
+  }
+
+  object queueRead extends Prim {
+    override val name: String = "queue-read"
+    override val minArgs: Int = 1
+    override val maxArgs: Int = 1
+
+    @checkTypeMismatch[Queue]
+    @checkArgumentMismatch
+    override def fn(ctxt: Ctxt): Either[PrimError, Ob] = {
+      val q = ctxt.argvec.elem.head.asInstanceOf[Queue]
+      if (q.isEmpty()) {
+        Left(QueueEmptyError)
+      }
+      Right(q.head)
+    }
+  }
+
+  object queuePDequeue extends Prim {
+    override val name: String = "queue-pat-dequeue"
+    override val minArgs: Int = 2
+    override val maxArgs: Int = 2
+
+    @checkArgumentMismatch
+    override def fn(ctxt: Ctxt): Either[PrimError, Ob] =
+      (ctxt.argvec.elem.head, ctxt.argvec.elem(1)) match {
+        case (q: Queue, pat: Tuple) =>
+          if (q.isEmpty) {
+            Left(QueueEmptyError)
+          } else if (pat == Tuple.NIL) {
+            Right(q.dequeue().get)
+          } else {
+            q.patternDequeue(pat)
+              .map(Right(_))
+              .getOrElse(Left(PatternMatchError))
+          }
+        case _ =>
+          Left(ArgumentMismatch(
+            "The first element should be a queue and second element should be a tuple"))
+      }
+  }
+
+  object queuePatRead extends Prim {
+    override val name: String = "queue-pat-read"
+    override val minArgs: Int = 2
+    override val maxArgs: Int = 2
+
+    @checkArgumentMismatch
+    override def fn(ctxt: Ctxt): Either[PrimError, Ob] =
+      (ctxt.argvec.elem.head, ctxt.argvec.elem(1)) match {
+        case (q: Queue, pat: Tuple) =>
+          if (q.isEmpty()) {
```

Comment:
queue-pat-read needs to return #absent for an empty queue, not PrimError.

Reply:
same as above